tests/controllers/oidc/oidc_provider_test.go

```diff
@@ -0,0 +1,382 @@
+	// register redirect endpoint. This endpoint will be called by the OIDC provider with a valid code.
+	mux.HandleFunc("/redirect", s.redirect)
+	s.server = httptest.NewServer(mux)
+}
+
+func (s *OIDCProviderSuite) TestOIDCAuthorizationCodeFlow() {
+	ctrl := gomock.NewController(s.T())
+
+	// create user and token
+	_, err := s.wranglerContext.Mgmt.User().Create(&apimgmtv3.User{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fakeUser,
+		},
+	})
+	assert.NoError(s.T(), err)
+	_, err = s.wranglerContext.Mgmt.Token().Create(&apimgmtv3.Token{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fakeToken,
+			Labels: map[string]string{
+				tokens.UserIDLabel: fakeUser,
+			},
+		},
+		AuthProvider: fakeAuthProvider,
+		Token:        fakeTokenValue,
+		UserID:       fakeUser,
+		Enabled:      ptr.To(true),
+	})
+	assert.NoError(s.T(), err)
+
+	// mock auth provider
+	mockProvider := providermocks.NewMockAuthProvider(ctrl)
+	mockProvider.EXPECT().IsDisabledProvider().Return(false, nil).AnyTimes()
+	providers.Providers[fakeAuthProvider] = mockProvider
+
+	// create OIDC client
+	oidcClient := &v3.OIDCClient{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fakeOIDCClient,
+			Annotations: map[string]string{
+				"foo": "bar",
+			},
+		},
+		Spec: v3.OIDCClientSpec{
+			RedirectURIs:                  []string{s.server.URL + "/redirect"},
+			TokenExpirationSeconds:        3600,
+			RefreshTokenExpirationSeconds: 36000,
+		},
+	}
+	_, err = s.wranglerContext.Mgmt.OIDCClient().Create(oidcClient)
+	assert.NoError(s.T(), err)
+
+	// wait for clientID and clientSecret to be created by the controller
+	assert.EventuallyWithT(s.T(), func(c *assert.CollectT) {
+		oidcClient, err = s.wranglerContext.Mgmt.OIDCClient().Get(oidcClient.Name, metav1.GetOptions{})
+		assert.NoError(c, err)
+		clientID = oidcClient.Status.ClientID
+		assert.NotEmpty(c, clientID)
+		secret, err := s.wranglerContext.Core.Secret().Get(clientSecretsNamespace, clientID, metav1.GetOptions{})
+		assert.NoError(s.T(), err)
+		clientSecret = string(secret.Data["client-secret-1"])
+		assert.NotEmpty(c, clientSecret)
+	}, duration, tick)
+
+	// set server URL
+	err = settings.ServerURL.Set(s.server.URL)
+	assert.NoError(s.T(), err)
+
+	// configure go oidc client
+	provider, err = goidc.NewProvider(s.ctx, s.server.URL+"/oidc")
+	assert.NoError(s.T(), err)
+
+	oauth2Config = oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		RedirectURL:  s.server.URL + "/redirect",
+		Endpoint:     provider.Endpoint(),
+		Scopes:       []string{goidc.ScopeOpenID, "profile", "offline_access"},
+	}
+
+	authURL := oauth2Config.AuthCodeURL(state, oauth2.S256ChallengeOption(fakeCodeVerifier))
+	req, err := http.NewRequest("GET", authURL, nil)
+	assert.NoError(s.T(), err)
+	req.Header.Set("Authorization", "Bearer "+fakeToken+":"+fakeTokenValue)
+	client := &http.Client{}
+	s.wg.Add(1)
+	res, err := client.Do(req)
```

Comment:
Do we need a custom client?

```suggestion
	s.wg.Add(1)
	res, err := http.DefaultClient.Do(req)
```